Implement _collection_ref to link objects to their NumberedObjectCollection parent

digvijay-y opened the pull request:
# Pull Request Checklist for MontePy

### Description

Added a ```_collection_ref ``` as a weak_ref which points the object back to its collection. 

# Problems
```deepcopy``` got messed up link "number already in use" another failed test due to surface conflicts. 
`_cell` pointed to wrong cell objects.
 
# Solution:
 Rebuild cache after `deepcopy` instead of overwriting `_collection_ref`.
 Added `_set_cell method` to `halfspace`

```
    100 objects -> 0.127 seconds
    500 objects -> 0.443 seconds
   1000 objects -> 0.937 seconds
   8000 objects -> 8.385 seconds
```

<details>

  ```Py
import time
import montepy


class Collection(montepy.Cells):
    """Use the real Cells collection for testing"""

    pass


def dump_stats(n):
    collection = Collection()
    start = time.time()
    for i in range(n):
        cell = montepy.Cell(f"{i+1} 0 -1")
        collection.append_renumber(cell)
    end = time.time()
    print(f"{n:>7d} objects -> {end - start:.3f} seconds")


if __name__ == "__main__":
    print("Performance test: append_renumber with _collection weakrefs")
    print("=" * 60)
    for n_cells in [100, 500, 1000, 8000]:
        dump_stats(n_cells)
    print("=" * 60)

  ```

</details>

Fixes #849 

---

### General Checklist

- [X] I have performed a self-review of my own code.
- [X] The code follows the standards outlined in the [development documentation](https://www.montepy.org/en/stable/dev_standards.html).
- [X] I have formatted my code with `black` version 25.
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable).

---

### LLM Disclosure

Were any large language models (LLM or "AI") used in to generate any of this code?

- [X] Yes
    - Model(s) used:
    -  Copilot
- [ ] No

<details open> 

<summary><h3>Documentation Checklist</h3></summary>

- [ ] I have documented all added classes and methods.
- [ ] For infrastructure updates, I have updated the developer's guide.
- [ ] For significant new features, I have added a section to the getting started guide.

</details>

---

<details>
<summary><h3>First-Time Contributor Checklist</h3></summary>

- [ ] If this is your first contribution, add yourself to `pyproject.toml` if you wish to do so.

</details>

---

### Additional Notes for Reviewers

Ensure that:

- [ ] This PR fully addresses and resolves the referenced issue(s).
- [ ] The submitted code is consistent with the merge checklist outlined [here](https://www.montepy.org/en/stable/dev_checklist.html#merge-checklist).
- [ ] The PR covers all relevant aspects according to the development guidelines.
- [ ] 100% coverage of the patch is achieved, or justification for a variance is given.


<!-- readthedocs-preview montepy start -->
----
📚 Documentation preview 📚: https://montepy--867.org.readthedocs.build/en/867/

<!-- readthedocs-preview montepy end -->

MicahGale commented on montepy/mcnp_object.py:
```diff
@@ -107,6 +107,7 @@ def __init__(
         except AttributeError:
             self._BLOCK_TYPE = montepy.input_parser.block_type.BlockType.DATA
         self._problem_ref = None
+        self._collection_ref = None
```
Good design. This should be for `NumberedMCNP_Object` though and not `mcnp_object`. The `_collection` reference only makes sense for numbered objects (for now).

MicahGale commented on montepy/numbered_mcnp_object.py:
```diff
@@ -14,22 +15,31 @@
+            if collection_type is not None:
+                collection = getattr(self._problem, collection_type.__name__.lower())
+            else:
+                collection = None
```
This should still raise an error eventually.

MicahGale commented on montepy/numbered_object_collection.py:
```diff
@@ -196,8 +197,18 @@ def link_to_problem(self, problem):
+        # after deepcopy/unpickling when the cache might be stale
+        self.__num_cache.clear()
+        for obj in self._objects:
+            self.__num_cache[obj.number] = obj
```
I'm not positive this is necessary. I thought it was handled through the MCNP_problem side.

MicahGale commented on montepy/numbered_object_collection.py:
```diff
@@ -196,8 +197,18 @@ def link_to_problem(self, problem):
+            if problem is not None:
+                existing_coll = obj._collection
+                if existing_coll is None or existing_coll._problem is not problem:
+                    obj._collection_ref = weakref.ref(self)
```
You should make a `_link_to_collection` method instead of doing this manually every time.

MicahGale reviewed (changes requested):
I didn't fully evaluate how this works, but I think there are some design changes needed first.

digvijay-y commented:
1. Moved `_collection_ref` to `Numbered_MCNP_Object`
2. Removed _set_cell methods
3. Fixed `_number_validator` to raise error
4.  Created _link_to_collection method
5. Removed cache rebuild
6.  Fixed try block error message

MicahGale commented on montepy/cell.py:
```diff
@@ -927,6 +927,10 @@ def num(obj):
+        # Clear weakrefs so the cloned cell isn't linked to original collection/problem
+        # This prevents number conflict checks against the original collection
+        result._collection_ref = None
+        result._problem_ref = None
```
This is still not the desired behavior in all cases. Generally they should be added to the same problem. I know that this can lead to a number collision in an edge case with hypothesis, that I haven't fixed yet, and I just ignore it with `rm -r .hypothesis`. I think that's a separate issue from this, and I need to open a bug report. Try it without these lines, and with purging the `.hypothesis` folder.

MicahGale commented on montepy/numbered_mcnp_object.py:
```diff
@@ -14,20 +15,30 @@
-        collection = getattr(self._problem, collection_type.__name__.lower())
+
+    # Only validate against collection if linked to a problem
+    if self._problem is not None:
```
You can skip a level of `if`s and just check `if self._collection is not None`.

MicahGale commented on montepy/numbered_mcnp_object.py:
```diff
@@ -14,20 +15,30 @@
+                raise TypeError(
+                    f"Could not find collection type for {type(self).__name__} in problem."
+                )
+
```
I think you can delete this whole branch. I don't see a case where `self._problem` is linked, but `self._collection` is not.

MicahGale commented on montepy/numbered_mcnp_object.py:
```diff
@@ -123,6 +135,24 @@ def _add_children_objs(self, problem):
 
+    @property
+    def _collection(self):
+        """Returns the parent collection this object belongs to, if any."""
```
You need to also reimplement `link_to_problem` here. It should at least:

1. call `super().link_to_problem`
2. grab the relevant collection from `problem`. See the logic above in `_validate_number`. 
3. create a `weakref.ref` to that collection (in `_collection_ref`). 

This will be the secondary way to update this ref, besides appending.

MicahGale replied:
Though this may lead to weird situations where objects are linked to problems without being in the collections. I'm going to belay this until I read further.

MicahGale commented on montepy/numbered_object_collection.py:
```diff
@@ -198,13 +199,38 @@ def link_to_problem(self, problem):
             return self._problem_ref()
         return None
 
+    def _link_to_collection(self, obj):
```
The reason being is objects should respect each other's private attributes. I am ok with calling each other's private functions as that in my mind is more marking it as internal use only.

MicahGale reviewed (changes requested):
This is a big improvement, and much more like what I was thinking. I have some more feedback, but it is not major.

Looking at the data you collected this does look like it is indeed linear (O(N)), so that's great that it is no longer O(N^2). I have been thinking more about how to present performance data recently. I think in general, rates should generally be used (e.g., objects/second) over just raw time. My last paper on this though did not use rates, so I'm working on this as well.

digvijay-y commented:
I can try data in object/second

MicahGale reviewed (changes requested):
A few of my old conversations were resolved a bit early, and still need to be addressed. Also one line of code is no longer covered by the test suite.

digvijay-y commented:
@MicahGale I guess there's some issue with CI, I tried Black . multiple times, but CI doesn't approve it.
